src/app/clusters/operational-state-server/operational-state-server.cpp

```
@@ -0,0 +1,317 @@
+        //
+        handlerContext.SetCommandHandled();
+
+        if (DataModel::Decode(handlerContext.mPayload, requestPayload) != CHIP_NO_ERROR)
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath,
+                                                     Protocols::InteractionModel::Status::InvalidCommand);
+            return;
+        }
+
+        func(handlerContext, requestPayload);
+    }
+}
+
+CHIP_ERROR OperationalStateServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    switch (aPath.mAttributeId)
+    {
+    case OperationalState::Attributes::OperationalStateList::Id: {
+        Delegate * delegate = OperationalState::GetOperationalStateDelegate(mEndpointId, mClusterId);
+        GenericOperationalState opState;
+        size_t index = 0;
+
+        VerifyOrReturnError(delegate != nullptr, CHIP_NO_ERROR, ChipLogError(NotSpecified, "Delegate is nullptr"));
+
+        // operational state list is not found
+        if (delegate->GetOperationalStateAtIndex(index, opState) == CHIP_ERROR_NOT_FOUND)
+        {
+            err = aEncoder.EncodeNull();
+        }
+        else
+        {
+            return aEncoder.EncodeList([&](const auto & encoder) -> CHIP_ERROR {
+                while (delegate->GetOperationalStateAtIndex(index, opState) != CHIP_ERROR_NOT_FOUND)
+                {
+                    ReturnErrorOnFailure(encoder.Encode(opState));
+                    index++;
+                }
+                return CHIP_NO_ERROR;
+            });
+        }
+    }
+    break;
+
+    case OperationalState::Attributes::OperationalState::Id: {
+
+        Delegate * delegate = OperationalState::GetOperationalStateDelegate(mEndpointId, mClusterId);
+        GenericOperationalState opState;
+        VerifyOrReturnError(delegate != nullptr, CHIP_NO_ERROR, ChipLogError(NotSpecified, "Delegate is nullptr"));
```

Again, this should probably return an error.  Same for the other attributes.